### PR TITLE
Hide refund button for non-captured orders

### LINKED ISSUE #2338: Incorrect/Contradictory UX when partially refunding authorized, uncaptured orders
**Describe the bug**
Right now, a partial refund on an authorized, uncaptured order results in the whole order being cancelled. While we've [made it somewhat more clear](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1481) to the merchant that this happens when they attempt that partial refund via notes, the refund button remains unchanged. This is bad UX, because there are contradictory messages to the user about how the feature works, and the most likely places for the user to look lead them to cancelling orders with the full expectation they will be partially refunded.

**To Reproduce**
1. Place a test order with Stripe when capturing is off.
2. Edit the order and click 'Refund'. 
3. The button will say 'Refund X.XX', instead of a correct message.

**Expected behavior**
I expect the button text/UI to match the actual behavior. Here are some options that would make for a predictable experience when partially refunding uncaptured orders:

* (Less work) When refunding uncaptured orders, options to refund anything but the whole amount are disabled and red text appears explaining that for partial refunds, the order must be set to 'processing' / captured first. 
* (More work) While the refund amount is under the order total, the button reads 'Capture payment and refund $X.XX via Stripe', and then that sequence is performed for the user. When the full amount is entered, it reads 'Cancel Order' or a similar message that communicates that the order will be cancelled, not refunded. 

**Screenshots**
Screenshot
*If you do not see the image, please [click here](https://d.pr/i/F9kejQ).*

**Additional context**
* This was suggested previously, but I couldn't find an actual issue for it: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/757#issuecomment-509107961
* 35138597-hc

To begin, I set this to 'Medium' priority. The reason is that businesses which authorize first, then capture are much less common than capture first; that said, I think that the risk of accidentally cancelled an order instead of partially refunding may lead to embarrassing and unprofessional situations for merchants. I don't know of any other UX flow where a control says it does one thing, and you have to read the order notes to know what other, different thing will happen when you hit the button.

### PR DESCRIPTION
**Pull request by EvanDotPro**
Currently, when merchants have the "capture later" setting enabled, a refund button is displayed for new orders which are authorized but not captured. However, this makes no sense because there's no such thing as refunding a charge that has not been made yet.

The current behavior is that if they do ANY refund on an uncaptured order, full or partial, it will simply cancel the order and release the pre-authorization, but the UI does a poor job of indicating that and the behavior is inherently unintuitive.

This change simply checks if the payment for the order has been authorized but not charged, and if so, hides the refund button and replaces it with a disabled "Refunding unavailable" button which includes a tooltip which explains why.

Fixes #2338.

## Testing instructions

- Check out branch
- Go to WooCommerce Settings -> Payments -> Stripe -> Settings
- Under Transaction preferences, enable `Issue an authorization on checkout, and capture later`, click save changes
- Place an order
- In the admin, go to WooCommerce -> Orders, click the order you just created
- In the second section that shows the items in the order, you should **not** see the "Refund" button, but instead see a disabled "Refunding unavailable" button with a tooltip that displays additional info:

Finally, disable the `Issue an authorization on checkout, and capture later`, place another order, and make sure the normal refund button is still displayed and functions properly.

### CODE REVIEW

**Comment by melek** [comment]
Hi @EvanDotPro, this looks really good!

image

I think this is a great solution!

I did a bit of smoke testing by manually cancelling / capturing the transaction. In both cases, the 'Refund' button appeared again like normal, and then I would get an error when trying to refund (I think the error is expected, but unsure if we can detect an out of sync order like this or not; perhaps that would be a new issue):

Markup 2022-09-07 at 11 03 33

I think the error with desynced orders is not a big deal here, but I'm not sure why the capture status changes whether the refund button is shown or not.

File: includes/admin/class-wc-stripe-settings-controller.php [hunk 1]
```diff
@@ -17,18 +17,45 @@ class WC_Stripe_Settings_Controller {
...
+
 	/**
 	 * Constructor
 	 *
 	 * @param WC_Stripe_Account $account Stripe account
 	 */
-	public function __construct( WC_Stripe_Account $account ) {
+	public function __construct( WC_Stripe_Account $account, WC_Stripe_Payment_Gateway $gateway ) {
 		$this->account = $account;
+		$this->gateway = $gateway;
```
Comment (wjrosa) [inline includes/admin/class-wc-stripe-settings-controller.php:34]: This allows us to mock the gateway and test the controller more effectively.

**Comment by wjrosa** [comment]
> Hi @EvanDotPro, this looks really good!
> 
> image
> 
> I think this is a great solution!
> 
> I did a bit of smoke testing by manually cancelling / capturing the transaction. In both cases, the 'Refund' button appeared again like normal, and then I would get an error when trying to refund (I think the error is expected, but unsure if we can detect an out of sync order like this or not; perhaps that would be a new issue):
> 
> Markup 2022-09-07 at 11 03 33
> 
> I think the error with desynced orders is not a big deal here, but I'm not sure why the capture status changes whether the refund button is shown or not.

Hi @melek ! I took over this PR. 

> I did a bit of smoke testing by manually cancelling / capturing the transaction. In both cases, the 'Refund' button appeared again like normal, and then I would get an error when trying to refund (I think the error is expected, but unsure if we can detect an out of sync order like this or not; perhaps that would be a new issue):

That's nice! But I'm not sure how we could handle this without requesting Stripe and checking for the payment intent history (if it was a "capture-later" type). So, I think it is best to handle this in a new issue.

> I think the error with desynced orders is not a big deal here, but I'm not sure why the capture status changes whether the refund button is shown or not.

Can you explain this a bit, please? Are you talking about the implementation?

**Review by mattallan — APPROVED** [review]
Thanks for taking over this PR @wjrosa! :100:

The changes in this PR look good to me. I tested the following cases:
- [x] With UPE enabled
   - [x] automatically captured orders
   - [x] uncaptured orders 
image
   - [x] fully captured orders 
   - [x] partial captured orders 👍 (could only refund the amount captured)
- [x] Same tests above with UPE disabled

**Pull request merged**